source/includes/write/gridfs.c

```diff
@@ -0,0 +1,150 @@
+        // start-create-bucket
+        mongoc_database_t *db = mongoc_client_get_database (client, "db");
+
+        bson_error_t error;
+        if (!mongoc_gridfs_bucket_new (db, NULL, NULL, &error)) {
+            fprintf (stderr, "Failed to create bucket: %s\n", error.message);
+        }
+        // end-create-bucket
+    }
+
+    {
+        // Creates a GridFS bucket and specifies a custom name
+        // start-create-custom-bucket
+        mongoc_database_t *db = mongoc_client_get_database (client, "db");
+        bson_t opts;
+        bson_init(&opts);
+        BSON_APPEND_UTF8 (&opts, "bucketName", "myCustomBucket");
+
+        bson_error_t error;
+        if (!mongoc_gridfs_bucket_new (db, &opts, NULL, &error)) {
+            fprintf (stderr, "Failed to create bucket: %s\n", error.message);
+        }
+        // end-create-custom-bucket
+    }
+
+    {
+        // Opens an upload stream for a given file name
+        // start-open-upload-stream
+        bson_error_t error;
+        mongoc_stream_t *upload_stream = mongoc_gridfs_bucket_open_upload_stream (bucket, "my_file", NULL, NULL, &error);
+        if (upload_stream == NULL) {
+            fprintf (stderr, "Failed to create upload stream: %s\n", error.message);
+        }
+        const char *data = "Data to store";
+        mongoc_stream_write (upload_stream, data, strlen(data), -1);
+
+        mongoc_stream_close (upload_stream);
+        mongoc_stream_destroy (upload_stream);
+        // end-open-upload-stream
+    }
+
+    {
+        // Uploads the contents of a stream to a GridFS file
+        // start-upload-from-stream
+        mongoc_stream_t *file_stream = mongoc_stream_file_new_for_path ("/path/to/input_file", O_RDONLY, 0);
+        
+        bson_error_t error;
+        if (!mongoc_gridfs_bucket_upload_from_stream (bucket, "new_file", file_stream, NULL, NULL, &error)) {
+            fprintf (stderr, "Failed to upload file: %s\n", error.message);
+        }
+
+        mongoc_stream_close (file_stream);
+        mongoc_stream_destroy (file_stream);
+        // end-upload-from-stream
+    }
+
+    {
+        // Retrieves information about GridFS files
+        // start-retrieve-file-info
+        mongoc_cursor_t *cursor = mongoc_gridfs_bucket_find(bucket, bson_new(), NULL);
+        const bson_t *file_doc;
+
+        while (mongoc_cursor_next(cursor, &file_doc)) {
+            char *json = bson_as_json(file_doc, NULL);
+            printf("%s\n", json);
+            bson_free(json);
+        }
+        
+        mongoc_cursor_destroy (cursor);
+        // end-retrieve-file-info
+    }
+
+    {
+        // Creates a download stream for a given file ID
+        // start-open-download-stream
+        char buf[512];
+        bson_oid_t oid;
+        bson_oid_init_from_string (&oid, "66fb1b8ea0f84a74ee099e71");
+        bson_value_t file_id;
+        file_id.value_type = BSON_TYPE_OID;
+        memcpy(&file_id.value.v_oid, &oid, sizeof(bson_oid_t));
```

Comment:
Note to tech reviewer: this seems like an overly complicated way to pass a file ID, so let me know if there's a better way (same with the next example)